### PR TITLE
games-emulation/rmg: new package, add 0.6.5, 9999 (Mupen64Plus Frontend)

### PR DESCRIPTION
**Pull request by orbea**
This adds the RMG Mupen64Plus frontend which includes several updated mupen64plus plugins as dynamic libraries for the optimal emulated N64 experience.

~~One caveat is that users who  have both musl and AMDGPU may experience segfaults with the vulkan `mupen64plus-video-parallel` plugin, this appears to be an issue in mesa.~~

https://github.com/Rosalie241/RMG/issues/219
https://gitlab.freedesktop.org/mesa/mesa/-/issues/10488

One trivial patch which has not been upstreamed yet has been applied to the live ebuild where the upstream has not yet responded.

https://github.com/Themaister/Granite/pull/129

And it would be nice to have advice on the licensing, the frontend itself is `GPL-3` where the plugins which are separate binaries are often `GPL-2` or sometimes a more permissive license. One exception is the `mupen64plus-video-angrylion` plugin which is released under the non-free `XMAME` license, but this is disabled by default and hidden behind the `angrylion` USE flag. It can be disabled unconditionally if this is problematic?

### CODE REVIEW

**Comment by orbea** [comment]
I updated the PR for rmg-0.5.8, the musl + vulkan segfault has been fixed.

**Comment by orbea** [comment]
I updated the PR for rmg-0.5.9.

**Comment by orbea** [comment]
Updated for `0.6.2`.

**Comment by carmiker** [comment]
If there are any concerns about merging this PR, I can attest to this being a quality application. It is the easiest and most accessible way for someone to play the N64 games they own on Linux, especially for those who are just casual emulator users and not deeply involved in the hobby.

**Comment by orbea** [comment]
The deal is that based on discussions in IRC Gentoo developers refuse to either consider this PR or reject it leaving us in limbo. Given the number of open PRs (829) this pattern is far from rare...

Its a small wonder Gentoo suffers from a lack of maintainers.

**Comment by juippis** [comment]
Well, @gentoo-bot has printed informational text about "new packages". Not sure what else can be said. Anyway, it's pretty clear you don't even want to become a maintainer anymore, so let's reduce the open PR count by one!

**Comment by orbea** [comment]
@juippis,

> Well, @gentoo-bot has printed informational text about "new packages". Not sure what else can be said. 

I don't use :guru and in the past adding new packages this wasn't a concern. I also don't want to use :guru when most of the packages I tried there in the past have been broken and the wiki has this disclaimer.

https://wiki.gentoo.org/wiki/Project:GURU#Disclaimer

> Anyway, it's pretty clear you don't even want to become a maintainer anymore

That is not true, I'm not sure why you think that? I'm just tired of being jerked around. I would be happy to contribute if this wasn't the case.

**Comment by juippis** [comment]
> 
> > Anyway, it's pretty clear you don't even want to become a maintainer anymore
> 
> That is not true, I'm not sure why you think that? I'm just tired of being jerked around. I would be happy to contribute if this wasn't the case.

For whatever reason I thought you had reassigned your packages in gentoo (my memory is hazy, maybe we talked about this) and I detected a "fed up" attitude from your previous message. Sorry for throwing you around again like this.

I'll take a look tomorrow.

**Comment by orbea** [comment]
> For whatever reason I thought you had reassigned your packages in gentoo (my memory is hazy, maybe we talked about this) 

I don't recall doing this, maybe you are confusing me with someone else?

> and I detected a "fed up" attitude from your previous message.

That is true that I am a bit "fed up", but that has nothing to do with your actions or responses and as I already stated it would be great if this could be resolved so that my efforts (And of others) don't just stagnate until the unforeseeable future. If there is anything I can do to help this, please let me know.

File: games-emulation/rmg/rmg-0.6.5.ebuild
Comment (juippis) [inline games-emulation/rmg/rmg-0.6.5.ebuild:56]: While I don't see anything exceptionally wrong with this, we prefer to use profile (use) masking for these. I'd mask the use flag on base profile, then unmask it on supported arches (amd64). Then again, do you expect to keyword this on other arches at all?

File: games-emulation/rmg/rmg-0.6.5.ebuild
Comment (juippis) [inline games-emulation/rmg/rmg-0.6.5.ebuild:38]:
```suggestion
	EGIT_REPO_URI="https://github.com/Rosalie241/${MY_PN}.git"
```

File: games-emulation/rmg/rmg-0.6.5.ebuild
Comment (juippis) [inline games-emulation/rmg/rmg-0.6.5.ebuild:69]: In proxy-maint we prefer the unconditionals before conditionals (minor style nitpick)
```suggestion
	dev-libs/hidapi:=
	dev-qt/qtbase:6[gui,widgets,vulkan]
	dev-qt/qtsvg:6
	media-libs/freetype:=
	media-libs/libpng:=
	media-libs/libsamplerate
	media-libs/libsdl2[haptic,joystick,sound,vulkan]
	media-libs/speexdsp
	sys-libs/zlib[minizip(+)]
	virtual/opengl
	rust? ( dev-libs/libusb:1 )
```
Oh yea and zlib doesn't "really" have a subslot.

File: games-emulation/rmg/metadata.xml
Comment (juippis) [inline games-emulation/rmg/metadata.xml:14]: I wonder if this particular use flag could be a bit more descriptive? `dynarec`? (tbh I'd rename all plugin use flags to `x-plugin`, but that's just me)

**Review by juippis — COMMENTED** [review]
I couldn't compile this initially due to:
```
In file included from /var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/3rdParty/mupen64plus-core/src/api/vidext.c:50:
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/3rdParty/mupen64plus-core/src/api/vidext_sdl2_compat.h:28:2: error: #error SDL is not build with OpenGL support. Try USE_GLES=1
   28 | #error SDL is not build with OpenGL support. Try USE_GLES=1
      |  ^~~~~
make: *** [/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/3rdParty/mupen64plus-core/projects/unix/Makefile:862: _obj/api/vidext.o] Error 1
```
After re-merging libsdl2 with `+opengl` in addition to the specified use flags, I got rid of that error but got this next:
```
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/UserInterface/Widget/Render/OGLWidget.cpp: In constructor 'UserInterface::Widget::OGLWidget::OGLWidget(QWidget*)':
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/UserInterface/Widget/Render/OGLWidget.cpp:34:46: error: invalid use of incomplete type 'class QOpenGLContext'
   34 |     this->openGLcontext = new QOpenGLContext();
      |                                              ^
In file included from /usr/include/qt6/QtGui/qsurface.h:9,
                 from /usr/include/qt6/QtGui/qwindow.h:16,
                 from /usr/include/qt6/QtGui/QWindow:1,
                 from /var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/UserInterface/Widget/Render/OGLWidget.hpp:15,    
                 from /var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/UserInterface/Widget/Render/OGLWidget.cpp:10:      
/usr/include/qt6/QtGui/qsurfaceformat.h:13:7: note: forward declaration of 'class QOpenGLContext'
   13 | class QOpenGLContext;
      |       ^~~~~~~~~~~~~~
```
It is most likely some missing Qt use flag dep thing, but I'm not too familiar with Qt so would you have any ideas what it's missing? Granted this is one of those that may ever appear in "tinderboxes" only and not in true desktops. I can try to emerge the package with plasma profile, `+X +wayland` turned on **if** you don't have clues either.

File: games-emulation/rmg/rmg-0.6.5.ebuild
Comment (orbea) [reply games-emulation/rmg/rmg-0.6.5.ebuild]: Thanks for explaining this, I wasn't sure what the best method was? I added the profile masking in a new commit, I believe I did this correctly?

File: games-emulation/rmg/rmg-0.6.5.ebuild
Comment (orbea) [reply games-emulation/rmg/rmg-0.6.5.ebuild]: Good catch, fixed.

File: games-emulation/rmg/rmg-0.6.5.ebuild
Comment (orbea) [reply games-emulation/rmg/rmg-0.6.5.ebuild]: Fixed.

File: games-emulation/rmg/metadata.xml
Comment (orbea) [reply games-emulation/rmg/metadata.xml]: Both work for me, I changed them,

**Comment by orbea** [comment]
@juippis Thank you very much for the review!

It seems I missed the `opengl` USE flag requirement for both libsdl2 and qtbase, the build issues should be fixed now.

File: games-emulation/rmg/metadata.xml
Comment (juippis) [reply games-emulation/rmg/metadata.xml]: You took my proposal a bit too literally :P I meant like:
 - angrylion-plugin
 - discord-plugin
 - dynarec-plugin
 - rust-plugin

Just so the use flag more accurately indicates to the user what it's about.

I can rename these while merging if you agree to it.

File: games-emulation/rmg/rmg-0.6.5.ebuild
Comment (juippis) [reply games-emulation/rmg/rmg-0.6.5.ebuild]: Yep, looks correct!

**Review by juippis — APPROVED** [review]
```
 * QA Notice: Files built without respecting CFLAGS have been detected
 *  Please include the following list of files in your report:
 * /usr/lib64/RMG/Plugin/Input/libmupen64plus_input_gca.so


 * QA Notice: Files built without respecting LDFLAGS have been detected
 *  Please include the following list of files in your report:
 * /usr/lib64/RMG/Plugin/Input/libmupen64plus_input_gca.so
```
would be nice if you looked into why that is, but it's not a requirement to get this PR merged. Let me know about the use flag thing.

**Comment by orbea** [comment]
> would be nice if you looked into why that is, but it's not a requirement to get this PR merged. 

I don't have those QA Notices, is there something I should enable to see them? That is the rust plugin so I am unsure if CFLAGS / LDFLAGS apply?

File: games-emulation/rmg/metadata.xml
Comment (orbea) [reply games-emulation/rmg/metadata.xml]: Oops, thanks for clarifying. I fixed it.

**Comment by juippis** [comment]
> I don't have those QA Notices, is there something I should enable to see them? That is the rust plugin so I am unsure if CFLAGS / LDFLAGS apply?

Ah, right of course. The rust problem. We can utilize `QA_FLAGS_IGNORED` to ignore it. I can add it while merging. I'll take a final look when I get back home tonight, thanks for fixing the issues!

**Comment by juippis** [comment]
*sigh* this worked last time and now I'm getting:
```
[172/203] /usr/bin/x86_64-pc-linux-gnu-g++ -DDRAG_DROP -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_DEBUG -DQT_WIDGETS_LIB -I/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5_build/Source/RMG/RMG_autogen/include -I/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5_build/Source/RMG -I/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG -I/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/.. -I/usr/include/SDL2 -I/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/3rdParty/imgui -isystem /usr/include/qt6/QtGui -isystem /usr/include/qt6 -isystem /usr/include/qt6/QtCore -isystem /usr/lib64/qt6/mkspecs/linux-g++ -isystem /usr/include/qt6/QtWidgets  -march=native -O2 -pipe -frecord-gcc-switches -std=gnu++20 -fPIC -MD -MT Source/RMG/CMakeFiles/RMG.dir/VidExt.cpp.o -MF Source/RMG/CMakeFiles/RMG.dir/VidExt.cpp.o.d -o Source/RMG/CMakeFiles/RMG.dir/VidExt.cpp.o -c /var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp
FAILED: Source/RMG/CMakeFiles/RMG.dir/VidExt.cpp.o
/usr/bin/x86_64-pc-linux-gnu-g++ -DDRAG_DROP -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_DEBUG -DQT_WIDGETS_LIB -I/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5_build/Source/RMG/RMG_autogen/include -I/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5_build/Source/RMG -I/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG -I/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/.. -I/usr/include/SDL2 -I/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/3rdParty/imgui -isystem /usr/include/qt6/QtGui -isystem /usr/include/qt6 -isystem /usr/include/qt6/QtCore -isystem /usr/lib64/qt6/mkspecs/linux-g++ -isystem /usr/include/qt6/QtWidgets  -march=native -O2 -pipe -frecord-gcc-switches -std=gnu++20 -fPIC -MD -MT Source/RMG/CMakeFiles/RMG.dir/VidExt.cpp.o -MF Source/RMG/CMakeFiles/RMG.dir/VidExt.cpp.o.d -o Source/RMG/CMakeFiles/RMG.dir/VidExt.cpp.o -c /var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:38:24: error: aggregate 'QVulkanInstance l_VulkanInstance' has incomplete type and cannot be defined
   38 | static QVulkanInstance l_VulkanInstance;
      |                        ^~~~~~~~~~~~~~~~
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:39:26: error: 'QVulkanExtension' was not declared in this scope
   39 | static QVulkanInfoVector<QVulkanExtension> l_VulkanExtensions;
      |                          ^~~~~~~~~~~~~~~~
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:39:26: error: 'QVulkanExtension' was not declared in this scope
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:39:26: error: 'QVulkanExtension' was not declared in this scope
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:39:26: error: 'QVulkanExtension' was not declared in this scope
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:39:26: error: 'QVulkanExtension' was not declared in this scope
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:39:26: error: 'QVulkanExtension' was not declared in this scope
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:39:26: error: 'QVulkanExtension' was not declared in this scope
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:39:26: error: 'QVulkanExtension' was not declared in this scope
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:39:8: error: 'QVulkanInfoVector' does not name a type
   39 | static QVulkanInfoVector<QVulkanExtension> l_VulkanExtensions;
      |        ^~~~~~~~~~~~~~~~~
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp: In function 'm64p_error VidExt_VK_GetSurface(void**, void*)':
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:407:33: error: 'VK_NULL_HANDLE' was not declared in this scope
  407 |     if ((VkInstance)Instance == VK_NULL_HANDLE)
      |                                 ^~~~~~~~~~~~~~
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:414:42: error: 'VK_NULL_HANDLE' was not declared in this scope
  414 |     if (l_VulkanInstance.vkInstance() == VK_NULL_HANDLE)
      |                                          ^~~~~~~~~~~~~~
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:426:51: error: incomplete type 'QVulkanInstance' used in nested name specifier
  426 |     VkSurfaceKHR vulkanSurface = QVulkanInstance::surfaceForWindow((*l_VulkanWidget));
      |                                                   ^~~~~~~~~~~~~~~~
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:427:26: error: 'VK_NULL_HANDLE' was not declared in this scope
  427 |     if (vulkanSurface == VK_NULL_HANDLE)
      |                          ^~~~~~~~~~~~~~
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp: In function 'm64p_error VidExt_VK_GetInstanceExtensions(const char***, uint32_t*)':
/var/tmp/portage/games-emulation/rmg-0.6.5/work/RMG-0.6.5/Source/RMG/VidExt.cpp:444:5: error: 'l_VulkanExtensions' was not declared in this scope; did you mean 'l_VulkanExtensionList'?
  444 |     l_VulkanExtensions = l_VulkanInstance.supportedExtensions();
      |     ^~~~~~~~~~~~~~~~~~
      |     l_VulkanExtensionList
```
any idea as to why? My qtbase:6 and libsdl2 has `vulkan` use flag.

**Comment by orbea** [comment]
> Ah, right of course. The rust problem. We can utilize QA_FLAGS_IGNORED to ignore it. I can add it while merging. I'll take a final look when I get back home tonight, thanks for fixing the issues!

Thanks!

> any idea as to why? My qtbase:6 and libsdl2 has vulkan use flag.

That is really strange, it almost looks like qtbase was built with with `USE=vulkan` while the vulkan-headers was not installed, but the qtbase ebuild has this in `DEPEND`.
```
gui? (
                vulkan? ( dev-util/vulkan-headers )
        )
```
FWIW I am using a default profile.

**Comment by orbea** [comment]
@juippis Thank you very much for reviewing and merging this, I greatly appreciate it!

**Pull request closed without merging**
